src/tds/bulk.c

```diff
@@ -138,7 +138,17 @@ tds_bcp_init(TDSSOCKET *tds, TDSBCPINFO *bcpinfo)
 		 * Or, see if the "lower" part is unused (and zeroed out) at this point, and just do one assignment.
 		 */
 		curcol->funcs = resinfo->columns[i]->funcs;
-		curcol->column_type = resinfo->columns[i]->column_type;
+		/* I believe the whole behaviour of figuring the column type out of column_type is flawed. It should
+		 * be based (at least partially!) on column_usertype (I haven't read documentation but that's from 
+		 * trials and error and the content of systypes table). The below is a dirty hack.
+		 * usertype = 2 > VARCHAR. 25 > NVARCHAR
+		 */
+		if ((resinfo->columns[i]->column_type == SYBCHAR) &&
```

Comment:
I think this part of the series makes AppVeyor failing pretty bad. I don't really like the same structure to have 2 semantics based on the usage.

Reply:
I'm not sure what you mean and what to do of ur review here... if you are questioning the enum, I agree (two enum for the same value is dirty). Do you propose to split in two? One Sybase enum and another MSSQL enum? I would agree, but maybe one thing at a time? I would also add an enum for the user type which isn't properly handled by FreeTDS as far as I can tell (I haven't fully checked on MSSQL, but it seems to be more of a Sybase issue)

Reply:
I think that this hunk should be removed, the function is just copying a structure, and it should not changing it while copying to simplify another part of the code. About the enum, yes, it became a bit a problem. They are not 2 enums, originally there were not overlapping, then code and versions got forked, so there are some common values, then MS only numbers and Sybase only numbers which, accidentally are now overlapping. The `column_type` (and actually to confuse even move there's a `on_server.column_type`) is the "wire" type which describe how the data is encoded in the protocol. There's also (as you discovered) a `column_usertype` which better describe the type at the user level. For instance the `timestamp` type and `binary` type has the same `column_type` but different `column_usertype`. I think in MS `varchar` and `nvarchar` has different `on_server.column_type` so there should be no problem there. About Sybase probably you are right.

Reply:
What part exactly would you want removed? If I remove that chunk, the bulk copy won't work.

I do agree that the historic led to that complicated situation.

Surely the code should be rewritten, like separate enum, or even separate functions for each flavor of TDS (I can see it is already done for some functions, but the principle could be expanded). I'm willing to give it a shot, but I don't think it should be on that specific PR. 

I would advocate for incremental approach here, and a rewrite later. Plus, that would give me time to get used to the FreeTDS code (Not that I should be the one rewriting it, but since I'm the one complaining, I'm willing to do it if necessary).

Reply:
I think the error should be fixed in another place. If code is not able to recreate the same type as it was it should be fixed in the code that declare the type, if the code is not able to insert that type of column it should be fixed in the code that insert data. As you said that field is already confusing, adding more meaning to it won't surely improving the situation.

The fact that a `column_type` (and not `on_server.column_type`) solves a problem is already an indication that something at some other level is broken.

Which specific data type are you not able to copy? Is it a copy from different databases (like from Sybase to MSSQL) or just Sybase to Sybase?

About a new enumeration. Do you have an idea of what it should represent? Should they values be independent on type representation? Should be more database "friendly" taking into consideration usertype but not, for instance, if the type is short or long (so having 2 different representation on the row)? Another solution would be possibly to remove it and use just the `on_server` version. On a similar issue there's a function called `tds_get_conversion_type` which `removes` nullable types. Another thing to consider is that dblib is returning a similar enumeration (but reduced and somehow simplified like XML will became a TEXT or not distinguishing between VARCHAR and NVARCHAR). I think this enumeration would be better to be addressed outside the scope of this PR. I would like to prepare a new release shortly, surely this enumeration change won't be there.

Reply:
The reason I fix here *at the moment* is that I'm not sure FreeTDS handle **usertype** correctly. I really don't want to embark on a too long journey and leave that bcp function broken atm. That's up to you. I can already check for a longer term fix (it was my plan anyways), but I would be sad to leave something which doesn't work for Sybase. Thinking about it, many graphical clients which rely on jTDS (and thus, FreeTDS) all display wrong information because of that. dbVisualizer, dBeaver, and probably all of them...

From my tests, it is impossible to guess the column type based on the `on_server` type. It's a surjective transformation where both char and varchar give a XSYBCHAR on the server. So if you start from XSYBCHAR, you are stuck.

I think we should have an enumeration for usertype. One for Sybase and one for MSSQL.

I cannot copy data from Sybase to Sybase (I now forgot my test case...). Maybe I'll continue this discussion on the issue I opened because this is getting quite cluttered now. I will describe my tests in detail there.

Reply:
Also, sorry for my slow replies. I'm unexpectedly busy at work and at home. Hope I can help more soon enough.